File: sql-insight-core/src/main/java/org/gongxuanzhang/sql/insight/core/engine/json/JsonEngine.java
/*
 * Copyright 2023 java-mysql  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.engine.json;

import com.alibaba.fastjson2.JSONObject;
import lombok.extern.slf4j.Slf4j;
import org.gongxuanzhang.sql.insight.core.command.dml.Update;
import org.gongxuanzhang.sql.insight.core.engine.storage.StorageEngine;
import org.gongxuanzhang.sql.insight.core.event.DropDatabaseEvent;
import org.gongxuanzhang.sql.insight.core.event.DropTableEvent;
import org.gongxuanzhang.sql.insight.core.event.InsightEvent;
import org.gongxuanzhang.sql.insight.core.event.MultipleEventListener;
import org.gongxuanzhang.sql.insight.core.exception.CreateTableException;
import org.gongxuanzhang.sql.insight.core.exception.InsertException;
import org.gongxuanzhang.sql.insight.core.exception.RuntimeIoException;
import org.gongxuanzhang.sql.insight.core.object.Column;
import org.gongxuanzhang.sql.insight.core.object.DataType;
import org.gongxuanzhang.sql.insight.core.object.InsertRow;
import org.gongxuanzhang.sql.insight.core.object.Row;
import org.gongxuanzhang.sql.insight.core.object.Table;
import org.gongxuanzhang.sql.insight.core.object.value.Value;
import org.gongxuanzhang.sql.insight.core.result.ExceptionResult;
import org.gongxuanzhang.sql.insight.core.result.MessageResult;
import org.gongxuanzhang.sql.insight.core.result.ResultInterface;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * @author gongxuanzhangmelt@gmail.com
 **/
@Slf4j
public class JsonEngine implements StorageEngine , MultipleEventListener {

    private final JsonIncrementKeyCounter counter = new JsonIncrementKeyCounter();

    private static final int MAX_PRIMARY_KEY = 10000;

    private static final int MIN_PRIMARY_KEY = 1;

    @Override
    public String getName() {
        return "json";
    }

    @Override
    public List<String> tableExtensions() {
        return Collections.singletonList("json");
    }

    @Override
    public void openTable(Table table) {
        if (!table.getIndexList().isEmpty()) {
            return;
        }
        JsonPkIndex jsonPkIndex = new JsonPkIndex(table);
        table.getIndexList().add(jsonPkIndex);
    }

    @Override
    public ResultInterface createTable(Table table) {
        File dbFolder = table.getDatabase().getDbFolder();
        if (table.getExt().getPrimaryKeyIndex() == -1) {
            throw new CreateTableException("engine json table must have a primary key");
        }
        if (table.getColumnList().get(table.getExt().getPrimaryKeyIndex()).getDataType().getType() != DataType.Type.INT) {
            throw new CreateTableException("engine json table primary key must int ");
        }
        File jsonFile = new File(dbFolder, table.getName() + ".json");
        try {
            if (!jsonFile.createNewFile()) {
                log.warn("create file {} fail", jsonFile.getName());
            }
            List<String> initContent = new ArrayList<>();
            for (int i = 0; i < MAX_PRIMARY_KEY; i++) {
                initContent.add("");
            }
            Files.write(jsonFile.toPath(), initContent);
            log.info("write {} json to {}", initContent.size(), jsonFile.toPath().toAbsolutePath());
            return new MessageResult(String.format("成功创建%s", table.getName()));
        } catch (IOException e) {
            return new ExceptionResult(e);
        }
    }

    @Override
    public ResultInterface truncateTable(Table table) {
        File jsonFile = JsonEngineSupport.getJsonFile(table);
        List<String> initContent = new ArrayList<>();
        for (int i = 0; i < MAX_PRIMARY_KEY; i++) {
            initContent.add("");
        }
        try {
            Files.write(jsonFile.toPath(), initContent);
            counter.reset(table);
            return new MessageResult();
        } catch (IOException e) {
            return new ExceptionResult(e);
        }
    }

    @Override
    public void insertRow(InsertRow row) {
        counter.dealAutoIncrement(row);
        JSONObject jsonObject = fullAllColumnRow(row);
        File jsonFile = JsonEngineSupport.getJsonFile(row.getTable());
        int insertPrimaryKey = JsonEngineSupport.getJsonInsertRowPrimaryKey(row.getTable(), jsonObject);
        if (MAX_PRIMARY_KEY < insertPrimaryKey || insertPrimaryKey < MIN_PRIMARY_KEY) {
            throw new InsertException("engine json primary key must between " + MIN_PRIMARY_KEY + " and " + MAX_PRIMARY_KEY);
        }
        try {
            List<String> lines = Files.readAllLines(jsonFile.toPath());
            String currentLine = lines.get(insertPrimaryKey);
            if (!currentLine.isEmpty()) {
                throw new InsertException(String.format("Duplicate entry '%s' for key 'PRIMARY'", insertPrimaryKey));
            }
            lines.set(insertPrimaryKey, jsonObject.toString());
            log.info("insert {} to table [{}] ", jsonObject, row.getTable().getName());
            Files.write(jsonFile.toPath(), lines);
        } catch (IOException e) {
            throw new RuntimeIoException(e);
        }
    }

    @Override
    public void update(Row oldRow, Update update) {
        File jsonFile = JsonEngineSupport.getJsonFile(update.getTable());
        final int rowId = (int) oldRow.getRowId();
        try {
            List<String> lines = Files.readAllLines(jsonFile.toPath());
            String line = lines.get(rowId);
            JSONObject jsonObject = JSONObject.parseObject(line);
            update.getUpdateField().forEach((colName, expression) -> {
                Value expressionValue = expression.getExpressionValue(oldRow);
                jsonObject.put(colName, expressionValue.getSource());
                String newLine = jsonObject.toString();
                lines.set(rowId, newLine);
                log.info("update {} to {} ", line, newLine);
            });
            Files.write(jsonFile.toPath(), lines);
        } catch (IOException e) {
            throw new RuntimeIoException(e);
        }
    }

    @Override
    public void delete(Row deletedRow) {
        long rowId = deletedRow.getRowId();
        File jsonFile = JsonEngineSupport.getJsonFile(deletedRow.belongTo());
        try {
            List<String> lines = Files.readAllLines(jsonFile.toPath());
            lines.set((int) rowId, "");
            Files.write(jsonFile.toPath(), lines);
        } catch (IOException e) {
            throw new RuntimeIoException(e);
        }
    }

    @Override
    public void refresh(Table table) {
        log.warn("The json engine dose not refresh manually because in update or delete already refresh");
    }


    private JSONObject fullAllColumnRow(InsertRow row) {
        Table table = row.getTable();
        JSONObject jsonObject = new JSONObject();
        table.getColumnList().forEach(col -> jsonObject.put(col.getName(), null));
        List<Column> insertColumns = row.getInsertColumns();
        List<Value> values = row.getValues();
        for (int i = 0; i < values.size(); i++) {
            jsonObject.put(insertColumns.get(i).getName(), values.get(i).getSource());
        }
        return jsonObject;
    }


    @Override
    public void onEvent(InsightEvent event) {
        if(event instanceof DropTableEvent){
            this.counter.reset(((DropTableEvent) event).getTable());
        }else if(event instanceof DropDatabaseEvent){
            this.counter.reset(((DropDatabaseEvent) event).getDatabase());
        }
    }

    @Override
    public List<Class<? extends InsightEvent>> listenEvent() {
        return Arrays.asList(DropTableEvent.class, DropDatabaseEvent.class);
    }
}

File: sql-insight-core/src/main/kotlin/tech/insight/core/engine/json/JsonEngine.kt
/*
 * Copyright 2023 sql-insight  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.insight.core.engine.json

import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import tech.insight.core.bean.DataType
import tech.insight.core.bean.InsertRow
import tech.insight.core.bean.Row
import tech.insight.core.bean.Table
import tech.insight.core.command.UpdateCommand
import tech.insight.core.engine.storage.StorageEngine
import tech.insight.core.event.DropDatabaseEvent
import tech.insight.core.event.DropTableEvent
import tech.insight.core.event.InsightEvent
import tech.insight.core.event.MultipleEventListener
import tech.insight.core.exception.CreateTableException
import tech.insight.core.exception.InsertException
import tech.insight.core.exception.TableDontOpenException
import tech.insight.core.extension.json
import tech.insight.core.extension.tree
import tech.insight.core.logging.Logging
import tech.insight.core.result.MessageResult
import tech.insight.core.result.ResultInterface
import java.io.File
import java.nio.file.Files
import java.util.concurrent.ConcurrentHashMap

/**
 * @author gongxuanzhangmelt@gmail.com
 */
class JsonEngine : Logging(), StorageEngine, MultipleEventListener {

    companion object {
        const val MIN_PRIMARY_KEY = 1
        const val MAX_PRIMARY_KEY = 10000
        val tableCache: MutableMap<Table, MutableList<String>> = ConcurrentHashMap()
    }

    private val counter = JsonIncrementKeyCounter
    override val name: String
        get() = "json"

    override fun tableExtensions(): List<String> {
        return listOf("json")
    }

    override fun openTable(table: Table) {
        if (table.indexList.isNotEmpty()) {
            return
        }
        val jsonPkIndex = JsonPkIndex(table)
        table.indexList.add(jsonPkIndex)
        tableCache.computeIfAbsent(table) {
            val jsonFile = JsonEngineSupport.getJsonFile(table)
            jsonFile.readLines().toMutableList()
        }

    }

    override fun createTable(table: Table): ResultInterface {
        val dbFolder: File = table.database.dbFolder
        if (table.ext.primaryKeyIndex == -1) {
            throw CreateTableException("engine json table must have a primary key")
        }
        if (table.columnList[table.ext.primaryKeyIndex].dataType != DataType.INT) {
            throw CreateTableException("engine json table primary key must int ")
        }
        val jsonFile = File(dbFolder, "${table.name}.json")
        if (!jsonFile.createNewFile()) {
            warn("create file {} fail", jsonFile.getName())
        }
        val initContent: MutableList<String> = ArrayList()
        for (i in 0 until MAX_PRIMARY_KEY) {
            initContent.add("")
        }
        Files.write(jsonFile.toPath(), initContent)
        info("write {} json to {}", initContent.size, jsonFile.toPath().toAbsolutePath())
        return MessageResult("success create table ${table.name}")
    }

    override fun truncateTable(table: Table): ResultInterface {
        val jsonFile = JsonEngineSupport.getJsonFile(table)
        val initContent: MutableList<String> = ArrayList()
        for (i in 0 until MAX_PRIMARY_KEY) {
            initContent.add("")
        }
        tableCache.remove(table)
        Files.write(jsonFile.toPath(), initContent)
        counter.reset(table)
        return MessageResult("success truncate table ${table.name}")
    }

    override fun insertRow(row: InsertRow) {
        counter.dealAutoIncrement(row)
        val jsonObject = fullAllColumnRow(row)
        val insertPrimaryKey = JsonEngineSupport.getJsonInsertRowPrimaryKey(row.belongTo(), jsonObject)
        check(insertPrimaryKey in MIN_PRIMARY_KEY..MAX_PRIMARY_KEY) {
            InsertException("engine json primary key must between $MIN_PRIMARY_KEY and $MAX_PRIMARY_KEY")
        }
        val lines = getLinesCache(row.belongTo())
        val currentLine = lines[insertPrimaryKey]
        check(currentLine.isEmpty()) {
            InsertException("Duplicate entry $insertPrimaryKey for key 'PRIMARY'")
        }
        lines[insertPrimaryKey] = jsonObject.toString()
        debug { "insert $jsonObject to table [${row.table.name}] " }
    }

    override fun update(oldRow: Row, update: UpdateCommand) {
        val lines = getLinesCache(update.table)
        val rowId = oldRow.rowId.toInt()
        val line = lines[rowId]
        val jsonNode = line.tree()
        update.updateField.forEach { (colName, expression) ->
            val expressionValue = expression.getExpressionValue(oldRow)
            jsonNode.put(colName, expressionValue.source)
            val newLine = jsonNode.json()
            lines[rowId] = newLine
            debug { "update $line to $newLine" }
        }
    }

    override fun delete(deletedRow: Row) {
        val rowId: Long = deletedRow.rowId
        val lines = getLinesCache(deletedRow.belongTo())
        lines[rowId.toInt()] = ""
    }

    override fun refresh(table: Table) {
        val lines = tableCache[table]
        Files.write(JsonEngineSupport.getJsonFile(table).toPath(), lines)
    }

    private fun getLinesCache(table: Table): MutableList<String> {
        return tableCache[table] ?: throw TableDontOpenException(table)
    }

    private fun fullAllColumnRow(row: InsertRow): ObjectNode {
        val jsonObject = jacksonObjectMapper().createObjectNode()
        val values = row.values
        row.table.columnList.forEachIndexed { index, column ->
            jsonObject.put(column.name, values[index].source)
        }
        return jsonObject
    }

    override fun onEvent(event: InsightEvent) {
        if (event is DropTableEvent) {
            counter.reset(event.table)
            tableCache.remove(event.table)
        } else if (event is DropDatabaseEvent) {
            counter.reset(event.database)
            tableCache.keys.filter { it.database == event.database }.forEach { tableCache.remove(it) }
        }
    }

    override fun listenEvent(): List<Class<out InsightEvent?>> {
        return listOf(DropTableEvent::class.java, DropDatabaseEvent::class.java)
    }


    /**
     * support put object
     */
    private fun ObjectNode.put(key: String, any: Any?) {
        if (any == null || any is Unit) {
            this.putNull(key)
            return
        }
        if (any is String) {
            this.put(key, any)
            return
        }
        if (any is Int) {
            this.put(key, any)
            return
        }
    }


}
